### ISSUE TITLE
Validator integration

### ISSUE THREAD

**Issue by florianrusch** [jlandowner/helm-chartsnap#44]
@jlandowner thanks for this amazing tool!

I was looking for exactly something like this, because I work in a project with multiple different developers, which all have different levels of experiences with helm and how to write and modify helm charts.

I'm just wondering, if it's somehow possible to also integrate a validator tool to validate the generated yaml code against the kubernetes api. Did you or someone else already thought about something like that? It would be really easy, if the snapshots would be stored as plain yaml files.

**Comment by florianrusch** [comment]
One possibility I found so far is to use the `--output-dir` option of the helm template command. One downside of this is that you only can execute one test, else the last one are overriding each other. And the diff is not the right one.
```shell
helm chartsnap -c my-chart/ -u -- --output-dir ./foo/test
 RUNS  Snapshot testing chart=my-chart/ values=
time=2023-12-23T13:44:58.450+01:00 level=INFO msg="loading helm output is done with warning"
WARN: failed to recognize a resource. snapshot as Unknown: 
---
object:
    apiVersion: helm-chartsnap.jlandowner.dev/v1alpha1
    kind: Unknown
    raw: |+
        wrote ./foo/my-chart/templates/cronjob.yaml


---
 FAIL  Snapshot does not match chart=my-chart/ values=
Expected to match
--- kind= name= line=1
  - object:
-     apiVersion: batch/v1
-     kind: CronJob
-     metadata:
		...
+     apiVersion: helm-chartsnap.jlandowner.dev/v1alpha1
+     kind: Unknown
+     raw: |+
+         wrote ./foo/my-chart/templates/cronjob.yaml
+ 
```

**Comment by jlandowner** [comment]
Hi @florianrusch, Thank you for raising!

Could you provide more details about the integration of a varidator tool?
> integrate a validator tool to validate the generated yaml code against the kubernetes api.

Basically, all that is needed for Helm is the chart and values. Therefore, for integration testing, I thought you can use the test values files to e.g. `helm upgrade --dry-run`, instead of using snapshot files.
This is one of the reasons why chartsnap uses the pure values.yaml file as test spec.

So it was not intended for the snapshot files to integrate with other tools for now.

Also some implementation reasons led to the current snapshot file format.
- To capture all the outputs of Helm template command, chartsnap uses [`CombinedOutput()`](https://pkg.go.dev/os/exec#Cmd.CombinedOutput), which captures both stdout and stderr.
- For handling dynamic values, it decodes the helm output as Kubernetes's struct [`Unstructured`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured).
  And when the decode fails, it is handled as a `kind: Unknown`, which is a custom kind in chartsnap (sorry but this has not been documented yet)
- To expedite development, the current code about snapshot files was reused from my previous project.

I'm open to supporting a better format and appriciate the suggestion 🙌
I would like to understand the plain yaml format you expected more clearly.

**Comment by florianrusch** [comment]
Hi, thanks for your reply!

I don't have a specific tool in mind, but for example kubectl validate would require the yaml resources to validate against the official schema: https://github.com/kubernetes-sigs/kubectl-validate/#usage

There are also a lot of other tools that would require yaml resources. You can find a good list here: https://github.com/HighwayofLife/kubernetes-validation-tools

So the idea would be to use helm-chartsnap to generate the resources in yaml format and do a first validation (snapshot test). Then we could reuse the yaml output as input for the other tools.

**Comment by jlandowner** [comment]
Thank you for more information!

As I mentioned above, chartsnap captures stdout and stderr and unknown output as `kind: Unknown`. Currently this seems to be a hurdle even if snapshot file is a standard YAML.

I will look into them and try to find how chartsnap can be integrated with them(including using test values files, not snapshot file)

**Comment by jlandowner** [comment]
@florianrusch Thanks for looking into my codes. The codes about snapshot files (`pkg/snap`) was reused from my previous project. It can take multiple snapshots in a single snapshot file so it requires snapshot id to recognize a snapshot in a file.
You seemed to find it in test code.
https://github.com/jlandowner/helm-chartsnap/blob/main/pkg/snap/__snapshots__/snapshot_test.snap#L15

But for chartsnap, a single snapshot(=a test values file, set of output of helm template) is stored in a single snapshot file so it does not so matter. It's just the name of a test values file(`default` if not) and you can find it in the first line of snapshot files.
https://github.com/jlandowner/helm-chartsnap/blob/main/example/app1/__snapshots__/default.snap#L1

**Comment by jlandowner** [comment]
Supported output as YAML in #85 😊

**Comment by jlandowner** [comment]
I found snapshot file extention should be changed  from `.snap` to `.yaml` in order to integrate with `kubectl-validate`

https://github.com/jlandowner/helm-chartsnap/issues/113#issuecomment-2110434846

**Comment by florianrusch** [comment]
Personally, I like the '.snap' file extensions. It makes clear, that these files are not normal k8s resource yaml files and that they contain snapshots.

Maybe we can think of adding an additional suffix like `.snap.yaml` or `.snapshot.yaml`.

WDYT?

**Comment by jlandowner** [comment]
I agree. In fact, I am using `.snap.yaml` in the screenshot. Also, since the base name of the snapshot file is the same as the values file, it would be confused.

**Comment by florianrusch** [comment]
@jlandowner is this still a todo?

**Comment by jlandowner** [comment]
Absolutely, but I couldn't work on it these days😓

**Comment by florianrusch** [comment]
Hey @jlandowner please let me know if I can help in any way. I've already tried changing the file extensions locally and it works, but I think we need to find a way to not introduce this as a breaking change, right? Also, the amount of test stuff that would need to be adjusted looked huge at first glance. 😕

**Comment by jlandowner** [comment]
Thanks, @florianrusch and sorry for my late response.
Too slow but I have been working on it. I opened the PR #157 and can you check it?

Now I'm preparing to fix the Unknown resource error.

**Comment by florianrusch** [comment]
Awesome! You are crazy 😄 I will try it out in our project and will give you feedback later 🙂 .
